File: Data Structures/python/fenwick_tree.py
class FenwickTree:
    def __init__(self, l: list):
        l.insert(0, 0);
        self.tree=l
        self.N=len(l)
        for i in range(0, self.N):
            parent=i+self.lsb(i)
            if(parent<self.N): self.tree[parent]+=self.tree[i]
    
    def lsb(self, i):
        return (i & -i)
    
    def __prefixSum(self, i):
        sum=0
        while(i!=0):
            sum+=self.tree[i]
            i&=~self.lsb(i)
        return sum
    
    def sum(self, i, j):
        assert i<=j
        return self.__prefixSum(j)-self.__prefixSum(i-1)
    
    def get(self, i):
        return self.sum(i, i)
    
    def add(self, i, val):
        while(i<self.N):
            self.tree[i]+=val
            i+=self.lsb(i)
    
    def set(self, i, val):
        self.add(i, val-self.get(i))

ft=FenwickTree([1, 2, 3, 4, 5, 6, 7])
for i in range(1, 8):
    print(ft.sum(1, i), end=" ")

File: Data Structures/cpp/fenwick_tree.cpp
#include<iostream>
#include<vector>

class FenwickTree{
    private:
        std::vector<long> tree;
        std::vector<long>::size_type N;

        int lsb(int i){
            return (i & -i);
        }

        long prefixSum(int i){
            long sum=0L;
            while(i!=0){
                sum+=tree[i];
                i&=~lsb(i);  //i-=lsb(i);
            }
            return sum;
        }
    
    public:
        FenwickTree(std::vector<long> values){
            values.insert(values.begin(), 0);
            N=values.size()+1;
            tree=values;
            for(int i=1; i<N; i++){
                int parent=i+lsb(i);
                if(parent<N) tree[parent]+=tree[i];
            }
        }

        long sum(int i, int j){
            if(i>j){
                std::cout<<"Make sure left<=right";
                return -1;
            }
            else
                return prefixSum(j)-prefixSum(i-1);
        }

        long get (int i){
            return sum(i, i);
        }

        void add(int i, long val){
            while(i<N){
                tree[i]+=val;
                i+=lsb(i);
            }
        }

        void set(int i, long val){
            add(i, val-sum(i, i));
        }

};

int lsb(int i){
    return (i & -i);
}

int main(){
    FenwickTree ft({1, 2, 3, 4, 5, 6, 7});
    std::cout<<ft.sum(1, 7)<<" ";
    return 0;
}